From 09e613a54d21e8dbee6789d458dec9e9f855118f Mon Sep 17 00:00:00 2001
From: Ash Charles <ashcharles@gmail.com>
Date: Tue, 15 Sep 2015 13:41:14 -0700
Subject: [PATCH 6/6] pepper: Read environment from EEPROM

The board identification EEPROM on Gumstix Pepper family boards can
contain environment variable settings.  Read and setup the environment
appropriately if this data is present.

Signed-off-by: Ash Charles <ashcharles@gmail.com>
---
 board/gumstix/pepper/board.c | 47 +++++++++++++++++++++++++++++++-------------
 board/gumstix/pepper/board.h |  2 +-
 include/configs/pepper.h     |  2 +-
 3 files changed, 35 insertions(+), 16 deletions(-)

diff --git a/board/gumstix/pepper/board.c b/board/gumstix/pepper/board.c
index d76c28b..e1dc90d 100644
--- a/board/gumstix/pepper/board.c
+++ b/board/gumstix/pepper/board.c
@@ -32,6 +32,20 @@
 
 DECLARE_GLOBAL_DATA_PTR;
 
+static int read_eeprom(struct pepper_board_id *header)
+{
+	if (i2c_probe(CONFIG_SYS_I2C_EEPROM_ADDR)) {
+		return -ENODEV;
+	}
+
+	if (i2c_read(CONFIG_SYS_I2C_EEPROM_ADDR, 0, 1, (uchar *)header,
+		sizeof(struct pepper_board_id))) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
 #ifdef CONFIG_SPL_BUILD
 #define OSC	(V_OSCK/1000000)
 
@@ -106,20 +120,6 @@ const struct ctrl_ioregs ioregs_ddr2 = {
 	.dt1ioctl		= MT47H128M16RT25E_IOCTRL_VALUE,
 };
 
-static int read_eeprom(struct pepper_board_id *header)
-{
-	if (i2c_probe(CONFIG_SYS_I2C_EEPROM_ADDR)) {
-		return -ENODEV;
-	}
-
-	if (i2c_read(CONFIG_SYS_I2C_EEPROM_ADDR, 0, 1, (uchar *)header,
-		sizeof(struct pepper_board_id))) {
-		return -EIO;
-	}
-
-	return 0;
-}
-
 const struct dpll_params *get_dpll_ddr_params(void)
 {
 	struct pepper_board_id header;
@@ -193,6 +193,25 @@ int board_init(void)
 	return 0;
 }
 
+#ifdef CONFIG_BOARD_LATE_INIT
+int board_late_init(void)
+{
+	struct pepper_board_id header;
+
+	if (read_eeprom(&header) < 0) {
+		puts("Could not get board ID.\n");
+		return 0;
+	}
+
+	if (header.content == 1) {
+		setenv(header.env_var, header.env_setting);
+	}
+
+	return 0;
+}
+#endif
+
+
 #if (defined(CONFIG_DRIVER_TI_CPSW) && !defined(CONFIG_SPL_BUILD)) || \
 	(defined(CONFIG_SPL_ETH_SUPPORT) && defined(CONFIG_SPL_BUILD))
 static struct ctrl_dev *cdev = (struct ctrl_dev *)CTRL_DEVICE_BASE;
diff --git a/board/gumstix/pepper/board.h b/board/gumstix/pepper/board.h
index a6df319..0a4b66c 100644
--- a/board/gumstix/pepper/board.h
+++ b/board/gumstix/pepper/board.h
@@ -18,7 +18,7 @@ struct pepper_board_id {
 	unsigned char content;
 	char fab_revision[8];
 	char env_var[16];
-	char en_setting[64];
+	char env_setting[64];
 };
 
 /*
diff --git a/include/configs/pepper.h b/include/configs/pepper.h
index 0bdcf22..f92916b 100644
--- a/include/configs/pepper.h
+++ b/include/configs/pepper.h
@@ -10,7 +10,7 @@
 #define CONFIG_MMC
 #include <configs/ti_am335x_common.h>
 
-#undef CONFIG_BOARD_LATE_INIT
+#define CONFIG_BOARD_LATE_INIT
 #undef CONFIG_SPL_OS_BOOT
 
 /* Clock defines */
-- 
2.1.4

